### PR TITLE
gh-118926: Deferred reference counting GC changes for free threading

### PR DESCRIPTION
**Pull request by Fidget-Spinner**
This PR mainly introduces GC changes to the free threading GC to support deferred reference counting in the future.

To get this to work, new stack references *must* immediately live on the stack, without any interfering `Py_DECREF` or escaping code between when we put them on the stack. This ensures they are visible to the GC.

This also NULLs out the rest of the stack because the GC scans the entire stack. The stack pointer may be inconsistent between escaping calls (including those to `Py_DECREF`) so we need to scan the whole stack.

This PR removes the temporary immortalization introduced previously in #117783. I wanted to do this in a separate PR, but the only way to test this properly is to remove that hack. So it has to be bundled in this PR.

Finally, this PR fixes a few bugs in steals and borrows. This was only caught by the GC changes, not by the debugger that I was working on. Since these are untestable without the GC changes, I bundled them in.

Temporary perf regressions (6 threads):
```
object_cfunction PASSED: 3.6x faster
cmodule_function FAILED: 2.1x faster (expected: LOAD_ATTR from dict)
generator PASSED: 3.7x faster
pymethod FAILED: 2.5x faster (expected: LOAD_ATTR from pytype_lookup)
pyfunction FAILED: 2.9x faster (expected: LOAD_GLOBAL)
module_function FAILED: 2.9x faster (expected: LOAD_ATTR from module dict)
load_string_const PASSED: 4.0x faster
load_tuple_const PASSED: 3.8x faster
create_closure FAILED: 2.3x faster (unsure why)
create_pyobject FAILED: 1.5x faster (expected: LOAD_GLOBAL)
```

* Issue: gh-118926

### CODE REVIEW

**Comment by Fidget-Spinner** [comment]
!buildbot nogil

**Comment by Fidget-Spinner** [comment]
!buildbot nogil

**Comment by Fidget-Spinner** [comment]
!buildbot nogil

File: Include/internal/pycore_opcode_metadata.h
Comment (colesbury) [inline Include/internal/pycore_opcode_metadata.h]: Should `_PyList_FromStackRefSteal` be added to the list of non-escaping functions?

File: Objects/object.c
Comment (colesbury) [inline Objects/object.c:2435]: 🎉

File: Python/frame.c
Comment (colesbury) [inline Python/frame.c]: I don't feel super strongly about this, but maybe we should push the locals and stack visiting into the GC code. That would avoid having to expose `_Py_visit_decref` and `_Py_visit_decref_unreachable` and keep any weirdness local to the GC.

In other words, something like:

```
...
Py_VISIT(_PyFrame_GetCode(frame));
return _PyGC_VisitFrameStack(frame, visit, arg);
```

**Review by colesbury — COMMENTED** [review]
This looks pretty good to me. A few comments below.

File: Include/internal/pycore_opcode_metadata.h
Comment (Fidget-Spinner) [reply Include/internal/pycore_opcode_metadata.h]: Yes.

**Comment by markshannon** [comment]
> This was only caught by the GC changes, not by the debugger that I was working on. 

OOI, do you know why this was?

File: Python/bytecodes.c
Comment (markshannon) [inline Python/bytecodes.c]: This doesn't seem right. `value` is a reference to an object, like any other item on the stack.

I don't think the concept of a pointer to a stack ref makes a lot of sense. Arrays, yes, pointers, no.

I'm a bit worried that the code generator does what you want here, as it is isn't what you say, or what it should do.

**Review by markshannon — COMMENTED** [review]
This doesn't look correct to me. The `PyStackRef *` annotations are incorrect. Those values are `PyStackRef`s not pointers.

We need a principled way of ensuring that the stack pointer is correct and visible to the GC when execution escapes from the interpreter. I don't think this PR fulfils that requirement. 

Why not use the approach prototyped in https://github.com/python/cpython/pull/119875?
(But, rather than add explicit `SAVE_SP -- LOAD_SP` pairs, wrap the calls in an `ESCAPING_CALL` macro)

File: Python/bytecodes.c
Comment (Fidget-Spinner) [reply Python/bytecodes.c]: Yeah I'm trying to say "give me a pointer instead of the actual thing". And it does do what I want, is that not intended?

**Comment by Fidget-Spinner** [comment]
> Why not use the approach prototyped in #119875? (But, rather than add explicit `SAVE_SP -- LOAD_SP` pairs, wrap the calls in an `ESCAPING_CALL` macro)

I would like to, but we'd need to wrap anything with Py_DECREF too, and that's too annoying.

**Comment by markshannon** [comment]
It doesn't matter what is "annoying". It needs to be correct.

`Py_DECREF` doesn't escape, it is `Py_Dealloc` that escapes. So the `ESCAPING_CALL` can be embedded in a interpreter-only version of the `Py_DECREF` macro.

**Comment by Fidget-Spinner** [comment]
> ` a interpreter-only version of the `Py_DECREF` macro.

Sorry I don't quite understand what you mean here. do you mean we can re-define Py_DECREF for the interpreter loop (via the code generator or otherwise)? If so that sounds good to me, I can work on that in this PR if you'd like.

**Comment by markshannon** [comment]
Something like:
```C
#define INTERPRETER_DECREF(op) \
    if (--op->ob_refcnt == 0) { \
        ESCAPING_CALL(Py_Dealloc(op)); \
    }
```
(plus debug and free-threading variants)

File: Python/bytecodes.c
Comment (markshannon) [reply Python/bytecodes.c]: But `value` is not a pointer to a reference, it is a reference.
`value[1]` is what you want.

Comment (markshannon) [reply Python/bytecodes.c]: Although `value[1]` is also undesirable as it isn't clear *why* you want a single item array.

Ideally, we should mark any escaping code and let the code generator spill the stack pointer.
We will need this information anyway, if we are to implement top-of-stack caching, as that will need to spill the cached values to memory across escaping calls.

**Comment by Fidget-Spinner** [comment]
Alright. I'll work on porting your PR over to the new branch.

File: Python/bytecodes.c
Comment (colesbury) [reply Python/bytecodes.c]: We want a pointer to a specific stack location here. It's important that the result is assigned to the stack (and therefore visible to the GC) before anything that might trigger a GC happens.

This is not related to spilling the stack pointer or making the *stack pointer* visible to the GC.

Comment (colesbury) [reply Python/bytecodes.c]: To expand on my above comment: A `_PyStackRef` only protects an object from destruction if it's part of the interpreter's evaluation stack or frame's local variables. A `_PyStackRef` that exists only as a local variables in C doesn't prevent object destruction. For example:

**BAD**:
```c
_PyStackRef tmp = PyStackRef_FromPyObjectNew(...);
Py_DECREF(...); // e.g., potentially trigger a GC
stack_pointer[-1] = tmp;
```

The problem is that if `tmp` is the only reference to some object, then the GC will free it because `tmp` is not visible to the GC.

**GOOD**:
```c
stack_pointer[-1] = PyStackRef_FromPyObjectNew(...);
Py_DECREF(...); // e.g., potentially trigger a GC
```

Ideally, `_PyStackRef` would only exist as part of `_PyInterpreterFrame.localsplus` and never as local variables in C, but that's awkward and hard to enforce in C.

Comment (markshannon) [reply Python/bytecodes.c]: I am aware of all of that.

We don't want a pointer to a specific stack location.
We want the GC to be able to find all the references on the stack.

If the code generator lacks the functionality you need, then the solution is to add that functionality, not to add hacky workarounds

If the current implementation hides values on the stack from the GC, then that's a problem, but unless we explicitly mark where the GC might run (i.e. escaping calls) then there is no way for the code generator to make sure that the GC can see whole stack without keeping it in memory all the time, and that's bad for performance.

Comment (Fidget-Spinner) [reply Python/bytecodes.c]: The current implementation scans the whole stack instead of just to stacktop. There is no hiding of values.

**Comment by Fidget-Spinner** [comment]
> > This was only caught by the GC changes, not by the debugger that I was working on. 
> 
> 
> 
> OOI, do you know why this was?

Because these values escape to PyObject, which we don't track.

File: Python/bytecodes.c
Comment (Fidget-Spinner) [reply Python/bytecodes.c]: I addressed the review by using the correct notation in the DSL.

Comment (colesbury) [reply Python/bytecodes.c]: @markshannon, how is `value[1]` appropriate here? There's no 1-element array here.

> We don't want a pointer to a specific stack location.

Why do you say that? We want a pointer to the location of where `value` (the result) is stored on the stack.

> If the current implementation hides values on the stack from the GC...

It's not hiding values *on the interpreter evaluation stack* from the GC. Values in C local variables are not visible to the GC.

> then there is no way for the code generator to make sure that the GC can see whole stack without keeping it in memory all the time...

What are you referring to here? The stack is currently always in memory.

Comment (colesbury) [reply Python/bytecodes.c]: Never mind about the `value[1]` notation -- I see how it's used now.

**Comment by Fidget-Spinner** [comment]
@markshannon do you have any other comments? This PR does not depend on spilling the stack pointer. Though that would be useful for other reasons, I would like for that to be a separate PR though.

**Comment by markshannon** [comment]
I don't like this change.
Having to add magical `[1]` suffices for reasons that aren't at all clear without knowing the internals of the nogil GC is going to be a major maintenance headache.

Looking at `inst(UNPACK_SEQUENCE_TWO_TUPLE, (unused/1, seq -- val1[1], val0[1]))` as a maintainer, I would want to clean it up to `inst(UNPACK_SEQUENCE_TWO_TUPLE, (unused/1, seq -- val1, val0))`. If I did that, the only clue as to what went wrong would be mysterious crashes in the nogil build.

`bytecodes.c` is used as input to two interpreter generators, the JIT generator, at least one abstract interpreter, and various metadata generators. It needs to be clear.
All necessary information should be expressed explicitly, and, ideally, be checkable.


And please don't mark comments as "resolved" when they aren't.

**Comment by markshannon** [comment]
>> We don't want a pointer to a specific stack location.

> Why do you say that? We want a pointer to the location of where value (the result) is stored on the stack.

Because it is the job of the code generators to manage the stack. We shouldn't be guessing where the code generator is physically storing a particular value.

> What are you referring to here? The stack is currently always in memory.

If that were the case, there wouldn't be a need for the `[1]` suffices. The top of the stack is kept in registers (C locals) between uops in tier 1, and will soon be kept in registers between uops in tier 2.

**Comment by Fidget-Spinner** [comment]
> And please don't mark comments as "resolved" when they aren't.

Sorry my bad, I thought they were.

Ok I kind of get the problem you're explaining. So we would need a solution that:

1. Is clear in the bytecode DSL.
2. Does not need exposing a pointer at all to the instruction body. So no `*val0 = thing`. Preferrably you want `val0 = thing` instead. This is to ensure compatibility with TOS caching when those are put in explicit registers.

In that case I propose the following solution:
1. A new annotation, either `write_to_stack val0` or `&val0`.
2. The code generator for tier 1 will see the annotation, and for every assignment to the value, it will also write it directly to the stack after, via parsing the instruction body.
E.g. if we have the following code:
```C
foo(bar1, -- write_to_stack val0) {
    val0 = op(bar1);
    Py_DECREF(bar1);
}
```
the codegen will produce for tier 1:
```C
TARGET(foo) {
    PyObject *bar1, *bar0;
    bar1 = stack_pointer[-1];
    val0 = op(bar1);
    stack_pointer[-1] = val0;
    Py_DECREF(bar1);
}
```
For tier 2 register version, it will produce:
```
TARGET(foo) {
    PyObject *reg1, *reg0;
    reg1 = stack_pointer[-1];
    reg0 = op(bar1);
    stack_pointer[-1] = reg0;
    Py_DECREF(reg1);
}
```

So both tier 1 and tier 2 are happy. Does that sound good to you?

**Comment by markshannon** [comment]
> A new annotation, either write_to_stack val0 or &val0 or flush val0.

There two problems with explicitly flushing values:
* It is that it isn't *why* the value needs to be flushed. 
* There is no way to check that the annotations are in the correct place.

Flushing cached values to the stack, or not, should be the job of the code generator(s).
The instruction/uop definitions should declare the semantics not the desired generated code, if possible.

It might be quite a lot of work, but I really think we should mark any escaping calls, so that the code generator knows what escapes. `DECREF_INPUTS` is already handled by the code generator, so no changes to the code should be necessary in `bytecodes.c` regarding reference count operations.

**Comment by Fidget-Spinner** [comment]
> It might be quite a lot of work, but I really think we should mark any escaping calls, so that the code generator knows what escapes. `DECREF_INPUTS` is already handled by the code generator, so no changes to the code should be necessary in `bytecodes.c` regarding reference count operations.

Ok I can port your old PR over to this one. I presume we're flushing the values to the stack before every escaping call?

**Comment by colesbury** [comment]
@markshannon, writing all values to the stack is going to introduce an unnecessary performance regression.

This is holding up the free-threaded work for what seems like small aesthetic complaints. If the concern is about maintainer confusion, we can add comments or otherwise document the dozen or so places where this is used.

**Comment by markshannon** [comment]
This is not just aesthetics.
Maintainability is important. By degrading maintainability, you are making more work for others. Particularly for my team.

You say that "writing all values to the stack will by slow". Only for those parts of the stack that are in memory.
Which is why it is important to leave the choice of which parts of the stack to spill to memory and which parts to spill to the stack up to the code generator, as it will differ for different tiers and different platforms.

> the dozen or so places where this is used.

There are a hundred or more places where execution can escape from the interpreter, not counting `DECREF`s which add hundreds more. It seems unlikely that this PR correctly identifies all cases where a garbage collection can occur and correctly spills all the necessary values to the stack memory.

Even if it is correct, it stills lays traps for the unwary.
For example, `_BINARY_OP` contains the code:
```
    DECREF_INPUTS();
    ERROR_IF(res_o == NULL, error);
    res = PyStackRef_FromPyObjectSteal(res_o);
```
If this were changed to
```
    res = PyStackRef_FromPyObjectSteal(res_o);
    DECREF_INPUTS();
    ERROR_IF(PyStackRef_IsNull(res), error);
```
then it would be unsafe, yet there would be no warning or error from any tools.

**Comment by colesbury** [comment]
The rewritten `_BINARY_OP` example is safe because it uses `PyStackRef_FromPyObjectSteal`. The concern is with calls to 
`PyStackRef_FromPyObjectNew`.

> There are a hundred or more places where execution can escape from the interpreter

We should not be thinking about this in terms of where calls escape from the interpreter. We should be thinking about this in terms of where `PyStackRef_FromPyObjectNew()` calls occur, and ensure that those are always written to the stack.

If you want to refactor `PyStackRef_FromPyObjectNew` so that it takes a pointer or similar that's fine.

**Comment by Fidget-Spinner** [comment]
> Ok I can port your old PR over to this one. I presume we're flushing the values to the stack before every escaping call?

Sorry I forgot that this may introduce another perf regression on the free-threaded build, even if I limit it just to that. So I'm putting a pause on this. I think we should discuss this on Wednesday.

**Comment by markshannon** [comment]
In general, stackrefs *must* be spilled to the in-memory stack around any escaping call.
To do this automatically, we need to:
1. Identify all escaping calls
2. Track assignments to output values
3. Raise an error if an escaping call occurs unless *all* or *none* of the output values have been assigned
4. Generate the spill around the call. 
   * If none of the output values have been defined, we just spill the stack pointer after popping the inputs.
   * If all of the output values have been defined, we save the outputs to memory and save the stack pointer before making the escaping call.

#### Identifying all escaping calls. 

We have a whitelist, all other calls are escaping. The list needs updating, but it's mostly right.

#### Track assignments to output values

Assignments are easy to spot `name = ...`. Then we need do is see if `name` is an output variable. Any assignments in nested code, or multiple assignments, should be treated as an error. 

It is easy enough to change the code move assignments out of branches, and don't want the code gen to have to do flow analysis.

#### Generating spills around calls.

Once we having identified the escaping call, we will need to walk backwards and forwards around the call to identify the whole statement.
Once we've done that we need to emit the spill before and the reload after (if any reload is needed)

**Comment by markshannon** [comment]
In case that sounds too expensive, don't worry. It shouldn't be. 

Spilling the stack pointer is cheap as registers often need to be spilled across calls anyway.
We need to save the output values to memory in most cases anyway, we are just doing it a little earlier.

Although escaping calls are common in `bytecodes.c`, there are not so common dynamically. Dynamically, only about 10% of instructions include an escaping call and for many of those the additional cost is negligible for the reasons given above.

**Comment by colesbury** [comment]
@markshannon, that is not what we discussed and agreed to yesterday. What we discussed was tracking `PyStackRef_FromPyObjectNew` calls from the code generator and spilling those immediately.

* We do not need or want to track stackrefs in general for the GC, just `PyStackRef_FromPyObjectNew`
* There's no point in looking for spilling calls and waiting to write the result of `PyStackRef_FromPyObjectNew` in multiple places.

**Comment by markshannon** [comment]
Tracking just `PyStackRef_FromPyObjectNew` will solve your immediate problem, but it doesn't help with a broader deferred reference counting implementation nor with top-of-stack caching.

I think what I proposed handles all the use cases, without a significant performance impact.
Even if it turns out the performance impact is too high, the additional analysis with help make any solution more robust.

> There's no point in looking for spilling calls and waiting to write the result of `PyStackRef_FromPyObjectNew` in multiple places.

Why would anything get written multiple times?

**Comment by Fidget-Spinner** [comment]
I think the original goal was to not block TOS caching and full deferred refcounting. With the latest changes, this is now true. The added goal of laying the foundations for the two should be left as another PR. The responsibility of this PR IMO, is to not burden stack caching and full deferred refcounting (which it should have achieved with the latest commit). Supporting more features is out of scope.

**Comment by colesbury** [comment]
@markshannon - I'm most concerned because I thought we were on the same page on this approach after our last meeting.

I don't think starting with this approach precludes future changes, like what you've outlined above. This PR is a bottleneck for a lot of the free-threading deferred reference counting work, so I'd appreciate it if we can figure out how to unblock it.

File: Python/executor_cases.c.h
Comment (markshannon) [inline Python/executor_cases.c.h:1439]: Don't emit anything if there are no values to flush.

File: Tools/cases_generator/generators_common.py
Comment (markshannon) [inline Tools/cases_generator/generators_common.py:192]: This is going to crash as soon as anyone writes anything but an assignment in front of `PyStackRef_FromPyObjectNew`.

We might want to insist that the result of `PyStackRef_FromPyObjectNew` is assigned to a variable, maybe only an output variable, but that should be an error not a crash.

File: Python/generated_cases.c.h
Comment (markshannon) [inline Python/generated_cases.c.h:1117]: Why is the same value being saved twice?

File: Tools/cases_generator/generators_common.py
Comment (markshannon) [inline Tools/cases_generator/generators_common.py:180]: I don't think that locally replacing `PyStackRef_FromPyObjectNew` is sufficient in general. It ensures that the references are in memory, but it doesn't ensure that the stack_pointer is saved, so the references might not be visible.

IIRC this is safe for the current FT implementation because you `NULL` out the unused part of the stack.
Could you add a comment explaining why this is safe for FT even though it doesn't appear to be.

File: Tools/cases_generator/generators_common.py
Comment (markshannon) [inline Tools/cases_generator/generators_common.py:194]: FYI, this sort of analysis should ideally be performed in the analysis phase, not in the code generation phase.
Don't worry about it for now.

File: Tools/cases_generator/generators_common.py
Comment (markshannon) [inline Tools/cases_generator/generators_common.py:203]: What happens if we haven't found a valid assignment?
That sounds like an error.

File: Tools/cases_generator/stack.py
Comment (markshannon) [inline Tools/cases_generator/stack.py:216]: It seems odd to be writing individual values to the stack. If any value on the stack needs to be in memory, don't they all?

Would it make more sense to use the `flush` method?

**Review by markshannon — CHANGES_REQUESTED** [review]
Having the spilling controlled by the code generator is a major improvement.
Thanks for doing that.

I have a few comments

File: Tools/cases_generator/stack.py
Comment (colesbury) [reply Tools/cases_generator/stack.py:216]: I'm looking at this since Ken is on vacation.

I think that the logic in this PR (after resolving merge conflicts with main) no longer works after https://github.com/python/cpython/pull/122286.

Before #122286, the tier1 generator had the outputs on the logical stack when `emit_tokens()` is called. Now, the inputs are still on the logical stack when `emit_tokens` is called and the outputs are only pushed after `emit_tokens()`.

I think `flush()` has the same issue: the logical stack still contains the inputs, not the outputs.

**Comment by colesbury** [comment]
@markshannon, I've split off the changes to the interpreter generator into a separate PR:

* https://github.com/python/cpython/pull/122748

I think that will make it easier to see the changes to the generated code. (And the merge conflicts were getting unwieldy.)

File: Tools/cases_generator/generators_common.py
Comment (colesbury) [reply Tools/cases_generator/generators_common.py:192]: I've updated the error handling. For the most part, `PyStackRef_FromPyObjectNew` is now required to be assigned to an output variable, but we also allow assignment to a dereferenced pointer due to usages like:

* `frame->localsplus[offset + i] = PyStackRef_FromPyObjectNew(o);` (in COPY_FREE_VARS)
* `*values++ = PyStackRef_FromPyObjectNew(items[i])` (in UNPACK_SEQUENCE_LIST)

File: Tools/cases_generator/generators_common.py
Comment (colesbury) [reply Tools/cases_generator/generators_common.py:180]: I added a comment

File: Tools/cases_generator/generators_common.py
Comment (colesbury) [reply Tools/cases_generator/generators_common.py:194]: I've moved this to `analyzer.py` since it required a number of changes anyways.

File: Python/generated_cases.c.h
Comment (colesbury) [reply Python/generated_cases.c.h:1117]: I updated the generator to avoid saving the same value twice. I got rid of the `#ifdef Py_GIL_DISABLED` guards because I think they would add too much complexity for no real gain.

File: Tools/cases_generator/generators_common.py
Comment (colesbury) [reply Tools/cases_generator/generators_common.py:203]: This now raises an `analysis_error`.

**Pull request closed without merging**
